### DISCUSSION TITLE
Show the Az segment when using either "az login" an "Connect-AzAccount"

### DISCUSSION THREAD

**Discussion by Ewerton** [JanDeDobbeleer/oh-my-posh#569] [General]
Tried to create an issue but the "Save Issue" is always disabled for me, so, here are a discussion...

### Description

The AZ segment works fine and show the correct powerline when I log using the [Azure CLI](https://docs.microsoft.com/pt-br/cli/azure/azure-cli-extensions-overview) and the [`az login`](https://docs.microsoft.com/pt-br/cli/azure/authenticate-azure-cli) command.
It does NOT work when I log into Azure using the [powershell module](https://docs.microsoft.com/pt-br/powershell/azure/install-az-ps?view=azps-5.7.0) and the [`Connect-AzAccount`](https://docs.microsoft.com/en-us/powershell/module/az.accounts/connect-azaccount?view=azps-5.7.0) command.

### Environment

- Oh my Posh version: 3.116.2
- Theme: custom (created based in the agnoster theme and with the "az" segment added)
- Operating System: Win10 (version 2004)
- Shell: PowerShell 7.1.3
- Terminal: Windows Terminal (1.7.572.0 Preview)

### Steps to Reproduce

1. Follow the [instructions](https://docs.microsoft.com/pt-br/powershell/azure/install-az-ps?view=azps-5.7.0) to install the Az Module for powershell 
2. Type [`Connect-AzAccount`](https://docs.microsoft.com/en-us/powershell/module/az.accounts/connect-azaccount?view=azps-5.7.0) to interactively connect to Azure.
3. Check that the powerline isn't showing the Subscription and Guid

**Expected behavior:** Show the powerline just like when I log using `az login`

**Actual behavior:** Powerline does not show informations about Azure Subscription and Guid

### Further Investigations
I'am not a Go developer, but looking at the [az segment code, right in the line 39](https://github.com/JanDeDobbeleer/oh-my-posh/blob/34dbdc6abbc49ec8b636f3a63fb308f69e1fe946/src/segment_az.go#L39) here in github it seems to me that the segment will be shown if the "az" text is present in the console, am I right? if so, it should be easy to add a condition to check if the "az" or "Connect-AzAccount" is present in the console to make it work

**Comment by JanDeDobbeleer** [comment]
It's not checking for the az text but that the az command is available. We can't query for the availability of Powershell cmdlets, nor execute these from go. Does it set an environment variable or something else after you run that command?

**Reply by Ewerton** [reply]
I've just compared all the environment variables before and after the login by `Connect-AzAccount` they are the same.

Otherwise, when the user is logged the `Get-AzContext` returns an object that represents the current tenant and subscription for the Azure, but as you said, we can't execute `Get-AzContext` from Go, right?

**Reply by tradiff** [reply]
I've played around with this a little bit.  For me, when using the Az PowerShell module, the information is stored in `~\.azure\AzureRmContext.json`.  @JanDeDobbeleer I can implement this if you're interested in reading from this file directly.  I don't know if this is ALWAYS the path or not, I didn't research that much.  Also things could get a little weird if you have both the AZ CLI, and the AZ PowerShell module set up.

**Reply by Ewerton** [reply]
Yeah Travis! `AzureRmContext.json` seems a good path to follow. By the way, I am using the MagicTooltips module right now, congratulations for the excelent work!

**Reply by Ewerton** [reply]
Bringing more information on this, the user can save the context by calling the [`Save-AzContext -Path <other path>`](https://docs.microsoft.com/en-us/powershell/module/az.accounts/save-azcontext?view=azps-5.7.0).

Note that when the user calls the `Save-AzContext` it does not change the location of the context file, it just saves a copy of the context file, the "Context" itself continues stored in `~\.azure\AzureRmContext.json`

Note that the `~\.azure\AzureRmContextSettings.json` is the file which stores the `ContextDirectory` and the `ContextFile`

**Reply by JanDeDobbeleer** [reply]
@TravisTX go for it!

**Reply by tradiff** [reply]
Bit of a good news/bad news situation.  I've got this working for reading the json files.  However that file doesn't get updated if the user switches subscriptions only for the current session, which is done using this command:
```sh
Select-AzSubscription -SubscriptionName 'DL-Development' -Scope Process
```

At that point, OMP and `Get-AzContext` will disagree with one another, which could lead to confusion or terrible accidents when someone takes action assuming they're in a different Subscription.

I see 3 options:
1. Scrap the whole idea
2. Add documentation explaining this segment doesn't apply to `-Scope Process`
3. Edit the PowerShell init script to execute `Get-AzContext` and pass the azure information in to OMP.

What are your thoughts @JanDeDobbeleer

**Reply by JanDeDobbeleer** [reply]
I feel like option 3 is the most reliable one. Is that command fast?

**Reply by tradiff** [reply]
There is another option that is more flexible for future scenarios, but puts a bit more burden on the user. We could change the Prompt function in the PowerShell init script to look for the presence of a `$global:PreOmp` script block.  If found, execute it before executing OMP.

That would allow the user to do something like this, coupled with an OMP `envvar` segment:

```pwsh
$global:PreOmp = {
    $env:AZ_SUBSCRIPTION_NAME = Get-AzContext | Select-Object -ExpandProperty "Subscription" | Select-Object -ExpandProperty "Name"
}
```

Edit: to answer your question, yes Get-AzContext is fast.  Seems to be pulling from memory.

**Reply by JanDeDobbeleer** [reply]
@TravisTX we already have `Set-PoshContext` for that no?

**Reply by tradiff** [reply]
Oh neat. I didn't realize that was a thing.  So @Ewerton you could add this to your PowerShell profile (tested on my end):

```pwsh
function Set-EnvVar {
    $env:AZ_SUBSCRIPTION_NAME = Get-AzContext | Select-Object -ExpandProperty "Subscription" | Select-Object -ExpandProperty "Name"
}

New-Alias -Name 'Set-PoshContext' -Value 'Set-EnvVar' -Scope Global
```

Along with an `envvar` segment in OMP.

@JanDeDobbeleer you still think we should call Get-AzContext from the PowerShell init script?  IMO this feels like a pretty niche feature for something that global.

**Reply by JanDeDobbeleer** [reply]
I wouldn't add it. Do we want to look for an env var in the AZ segment which people can then populate using this functionality?

**Reply by Ewerton** [reply]
@TravisTX I tested as you requested, here are what I done:
- Added the ENVVAR segment to my `oh-my-posh-custom.omp.json`
- Changed my profile, before everything I put the following:
`function Set-EnvVar {
    $env:AZ_SUBSCRIPTION_NAME = Get-AzContext | Select-Object -ExpandProperty "Subscription" | Select-Object -ExpandProperty "Name"
}`

     `New-Alias -Name 'Set-PoshContext' -Value 'Set-EnvVar' -Scope Global -Force`
_I just added the -Force param to the New-Alias to prevent errors when I reload my profile using `. $profile`_

- Logged to my Azure using `Connect-AzAccount` and the ENVVAR segment shows the correct subscription name
- Changed the Subscription using `Select-AzSubscription -Subscription "MySubscription" -Scope Process` and the ENVVAR segment shows the correct subscription name
- Changed the Subscription using `Select-AzContext 'MySubscription (XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX) - YYYYYYYY-YYYY-YYYY-YYYY-YYYYYYYYYYYY - email@email.com' -Scope Process` and the ENVVAR segment shows the correct subscription name
- Disconnected from my account using `Disconnect-AzAccount` and the ENVVAR segment doesn't shows anymore.

So, for me, everything is working fine.

**Comment by tradiff** [comment]
Sorry for the wait, I had some stuff come up. I've submitted #587 for review, which will close this issue.  @Ewerton once this PR gets completed, you will be able to use the az segment instead of the envvar segment.  It won't have any big difference for you, other than being "more official" 😁

**Reply by Ewerton** [reply]
Thank you man! As soon the new feature be approve I will upgrade and test it.

**Comment by Ewerton** [comment]
Hey Friends, I've just update to 3.129.0 and did some tests: 
**Using the Azure CLI**
- `az-login` - OK
- `az account -set --subscription <my_subscription>` - OK
- `az logout` - OK

**Using the Powershell cmdlets for Azure**
- `Connect-AzAccount` - OK
- `Select-AzSubscription -Subscription <my_subscription>` - OK
- `Disconnect-AzAccount` - NOT OK The powerline stills showing my last selected subscription for subsequent commands

**Additional Info**
My Az Segment is configured as follows:
```
{
    "type": "az",
    "style": "powerline",
    "powerline_symbol": "\uE0B0",
    "foreground": "#000000",
    "background": "#9ec3f0",
    "properties": {
	"display_id": false,
	"display_name": true,
	"info_separator": " @ ",
	"prefix": " \uFD03 " 
     }
}
```
Dont know if this can cause any problems but note that I opt to not show the subscription ID (`display_id = false`) and I put spaces between the "prefix" icon.

**Reply by tradiff** [reply]
I noticed this edge case too. It's because it's setting an environment variable, and that environment variable still exists.  It should eventually clear out (maybe on restarting powershell, if not, then definitely on restarting your computer).  It didn't seem worth the extra cycles to clear it, since it's code that runs for every powershell user (regardless of whether they use or have ever used the az powershell module)

**Reply by Ewerton** [reply]
Yep, I can confirm. The env vars still there after a `Disconnect-AzAccount`  (`Get-ChildItem Env:*AZ_*`)
I think that if the powerline keeps showing those information after the users logout, the users can think they still connected to Azure when, in fact, they aren't.

**Reply by tradiff** [reply]
Right.  It's an edge case. The fix is a trade-off I didn't find to be worth it, since the the risk is negligible. (You think you're connected to Azure, but you aren't, any Azure commands will just fail).

I can add it if Jan disagrees on the tradeoff.  The change would be to change the PowerShell init script to clear those environment variables if they're set and we're not connected to Azure.  It's just that it's code running for every PowerShell user, so that's why I opted not to do it.

**Reply by JanDeDobbeleer** [reply]
@TravisTX that would be in the catch statement or else, right? Works for me, not a heavyweight thing to do.
